lib/services/types.proto

```diff
@@ -975,3 +985,86 @@ message TunnelConnectionSpecV2 {
     // Type is the type of reverse tunnel, either proxy or node.
     string Type = 4 [ (gogoproto.jsontag) = "type", (gogoproto.casttype) = "TunnelType" ];
 }
+
+// SemaphoreFilter encodes semaphore filtering params.
+message SemaphoreFilter {
+    // SemaphoreKind is the kind of the semaphore.
+    string SemaphoreKind = 1 [ (gogoproto.jsontag) = "kind" ];
+    // SemaphoreName is the name of the semaphore.
+    string SemaphoreName = 2 [ (gogoproto.jsontag) = "name" ];
+}
+
+// AcquireSemaphoreParams holds semaphore lease acquisition parameters.
+message AcquireSemaphoreParams {
+    // SemaphoreKind is the kind of the semaphore.
+    string SemaphoreKind = 1 [ (gogoproto.jsontag) = "kind" ];
+    // SemaphoreName is the name of the semaphore.
+    string SemaphoreName = 2 [ (gogoproto.jsontag) = "name" ];
+    // MaxLeases is the maximum number of concurrent leases.  If acquisition
+    // would cause more than MaxLeases to exist, acquisition must fail.
+    int64 MaxLeases = 3 [ (gogoproto.jsontag) = "max_resources" ];
+    // LeaseID uniquely identifies the lease.
+    string LeaseID = 4 [ (gogoproto.jsontag) = "lease_id" ];
+    // Expires is the time at which this lease expires.
+    google.protobuf.Timestamp Expires = 5 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "expires"
+    ];
+    // Holder identifies the entitiy holding the lease.
+    string Holder = 6 [ (gogoproto.jsontag) = "holder" ];
+}
+
+// SemaphoreLease represents lease acquired for semaphore
+message SemaphoreLease {
+    // SemaphoreKind is the kind of the semaphore.
+    string SemaphoreKind = 1 [ (gogoproto.jsontag) = "kind" ];
+    // SemaphoreName is the name of the semaphore.
+    string SemaphoreName = 2 [ (gogoproto.jsontag) = "name" ];
+    // LeaseID uniquely identifies this lease.
+    string LeaseID = 3 [ (gogoproto.jsontag) = "lease_id" ];
+    // Expires is the time at which this lease expires.
+    google.protobuf.Timestamp Expires = 5 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "expires"
+    ];
+}
+
+// SemaphoreLeaseRef identifies an existent lease.
+message SemaphoreLeaseRef {
```

Comment:
`SemaphoreLease` specifies the name and kind of the semaphore, which `SemaphoreLeaseRef` does not need to store since `SemaphoreLeaseRef` is stored within its semaphore.  Meanwhile, `SemaphoreLeaseRef` stores `Holder`, which `SemaphoreLease` does not need to store since `Holder` is basically just debug info.

Generally, the two types have different requirements, so it seemed better to define them separately from the beginning.

Reply:
I guess it's a tradeoff between minimizing storage (purpose-built minimal objects) and schema simplicity (reusable objects where some fields are occasionally unused).
I like the latter because both `SemaphoreLease` and `SemaphoreLeaseRef` represent the same logical "thing".
But if you and @russjones both prefer the current approach, I'm OK with it.